moveit_ros/moveit_servo/src/servo_calcs.cpp

```diff
@@ -758,58 +758,25 @@ double ServoCalcs::velocityScalingFactorForSingularity(const Eigen::VectorXd& co
 
 void ServoCalcs::enforceVelLimits(Eigen::ArrayXd& delta_theta)
 {
+  // Convert to joint angle velocities for checking and applying joint specific velocity limits.
   Eigen::ArrayXd velocity = delta_theta / parameters_.publish_period;
 
   std::size_t joint_delta_index = 0;
-  // Track the smallest velocity scaling factor required, across all joints
-  double velocity_limit_scaling_factor = 1;
-
-  for (auto joint : joint_model_group_->getActiveJointModels())
+  double velocity_scaling_factor = 1;
+  for (const moveit::core::JointModel* joint : joint_model_group_->getActiveJointModels())
   {
-    // Some joints do not have bounds defined
-    const auto bounds = joint->getVariableBounds(joint->getName());
-
+    const auto& bounds = joint->getVariableBounds(joint->getName());
     if (bounds.velocity_bounded_)
     {
-      velocity(joint_delta_index) = delta_theta(joint_delta_index) / parameters_.publish_period;
-
-      bool clip_velocity = false;
-      double velocity_limit = 0.0;
-      if (velocity(joint_delta_index) < bounds.min_velocity_)
-      {
-        clip_velocity = true;
-        velocity_limit = bounds.min_velocity_;
-      }
-      else if (velocity(joint_delta_index) > bounds.max_velocity_)
-      {
-        clip_velocity = true;
-        velocity_limit = bounds.max_velocity_;
-      }
-
-      // Apply velocity bounds
-      if (clip_velocity)
-      {
-        const double scaling_factor =
-            fabs(velocity_limit * parameters_.publish_period) / fabs(delta_theta(joint_delta_index));
-
-        // Store the scaling factor if it's the smallest yet
-        if (scaling_factor < velocity_limit_scaling_factor)
-          velocity_limit_scaling_factor = scaling_factor;
-      }
+      const auto unbounded_velocity = velocity(joint_delta_index);
+      velocity(joint_delta_index) = std::min(std::max(unbounded_velocity, bounds.min_velocity_), bounds.max_velocity_);
+      velocity_scaling_factor = std::min(velocity_scaling_factor, unbounded_velocity / velocity(joint_delta_index));
```

Comment:
Commenting on my own mistakes here, but I think I got at least this wrong. The line above (772), first clamps the velocities to a range of `[min_velocity, max_velocity]`. That is correct. However, when calculating the scaling factor, the formula is definitely wrong and should be `velocity(joint_delta_index) / unbounded_velocity` instead.

Now, even with that fix, it seems to be possible to construct a situation where the `velocity_scaling_factor` ends up being nonsensical. For instance: if we have a set of joint velocities `[-1, -3, 2, 3, 4, 5]` and `max_velocity` of 2 and `min_velocity` of 1, then we would first clamp the velocities down to a set of `[1, 1, 2, 2, 2, 2]` and then get the minimum scaling factor of `-1` which would switch the directions the joints are moving. Now, I'm not entirely sure if this kind of scenario is possible with the way velocity limits are constructed. @AndyZe is it possible to have both the max and min velocities to be positive or are they guaranteed to always have opposing signs?

Reply:
I think there is only a "velocity_limit" in the URDF. It is always a positive value. Hence, min and max are symmetric. Will try to find a reference to confirm.

Reply:
This [page](http://wiki.ros.org/urdf/XML/joint) and this [page](http://wiki.ros.org/pr2_controller_manager/safety_limits) confirm that the velocity limit from the URDF is always given as a positive value. So the limits are (-velocity_limit, +velocity_limit).